tools/settings.py

```diff
@@ -85,12 +86,12 @@
 
 
 class SettingsManager:
-  attrs = {}
-  types = {}
-  allowed_settings = set()
-  legacy_settings = {}
-  alt_names = {}
-  internal_settings = set()
+  attrs: Dict[str, Any] = {}
+  types: Dict[str, Any] = {}
+  allowed_settings: Set[str] = set()
+  legacy_settings: Dict[str, tuple] = {}
+  alt_names: Dict[str, str] = {}
+  internal_settings: Set[str] = set()
```

Comment:
Do you feel this helps? While in general I agree static types are a good idea in large complex codebases, I'm not sure `emcc.py` is that big or that complex. Do you think this would catch bugs or help refactoring?

Reply:
I think it will do both yes.   It already found at least one real bug. 

Also, the types are often implicit so it shouldn't be too invasive.   

We can also do it incrementally, without investing a whole to up front.

Reply:
Fair enough... it looks odd to me but probably I just need to get used to looking at it.

We should probably get consensus on a wide level though before moving forward here. But I'm not opposed myself.

Reply:
Yeah, I agree with this. I think it probably makes sense to do but I'd like to know more about how we should use it and how we expect to add it to the codebase over time, and what the long-term goal should be.

Reply:
OK, I added some docs about this (and coding style and linting in general).